SerialPrograms/Source/PokemonSV/Inference/Map/PokemonSV_FastTravelDetector.cpp

```diff
@@ -0,0 +1,128 @@
+/*  Fast Travel Detector
+ *
+ *  From: https://github.com/PokemonAutomation/Arduino-Source
+ *
+ */
+
+#include "Common/Cpp/Containers/FixedLimitVector.tpp"
+#include "CommonFramework/ImageMatch/WaterfillTemplateMatcher.h"
+#include "CommonFramework/ImageTypes/ImageViewRGB32.h"
+#include "CommonFramework/ImageTools/WaterfillUtilities.h"
+#include "CommonFramework/VideoPipeline/VideoOverlayScopes.h"
+#include "Kernels/Waterfill/Kernels_Waterfill_Types.h"
+#include "PokemonSV_FastTravelDetector.h"
+
+// #include <iostream>
+//using std::cout;
+//using std::endl;
+
+namespace PokemonAutomation{
+namespace NintendoSwitch{
+namespace PokemonSV{
+
+ImageFloatBox MINIMAP_AREA(0.815, 0.680, 0.180, 0.310);
+
+
+class FastTravelMatcher : public ImageMatch::WaterfillTemplateMatcher{
+public:
+    FastTravelMatcher() : WaterfillTemplateMatcher(
+        "PokemonSV/Map/FastTravelIcon-Template.png", Color(0,0,0), Color(255, 255, 255), 450
+    ) {
+        m_aspect_ratio_lower = 0.9;
+        m_aspect_ratio_upper = 1.2;
+        m_area_ratio_lower = 0.6;
+        m_area_ratio_upper = 1.1;
+    }
+
+    static const ImageMatch::WaterfillTemplateMatcher& instance() {
+        static FastTravelMatcher matcher;
+        return matcher;
+    }
+};
+
+
+FastTravelDetector::~FastTravelDetector() = default;
+
+FastTravelDetector::FastTravelDetector(Color color, const ImageFloatBox& box)
+    : m_color(color)
+    , m_box(box)
+{}
+
+void FastTravelDetector::make_overlays(VideoOverlaySet& items) const{
+    items.add(m_color, m_box);
+}
+
+bool FastTravelDetector::detect(const ImageViewRGB32& screen) const{
+    std::vector<ImageFloatBox> hits = detect_all(screen);
+    return !hits.empty();
+}
+
+std::vector<ImageFloatBox> FastTravelDetector::detect_all(const ImageViewRGB32& screen) const{
+    const std::vector<std::pair<uint32_t, uint32_t>> filters = {
+        {combine_rgb(0, 0, 80), combine_rgb(140, 140, 255)},
+        {combine_rgb(5, 100, 200), combine_rgb(60, 135, 255)},
+    };
+
+    const double min_object_size = 400.0;
+    // need to keep RMSD threshold below 100, else will false positive the blue dots at blueberry academy
+    const double rmsd_threshold = 80.0;
+
+    const double screen_rel_size = (screen.height() / 1080.0);
```

Comment:
Has this been tested at 720p?

Reply:
Good point. I didn't. It does fail under certain circumstances at 720p (mostly at Blueberry academy). I'll try to see if I can fix it.

Reply:
I tried adjusting my detector to also work at 720p at the blueberry academy. With my current technique, this isn't possible without also increasing false positives. It works approximately 75-80% of the time depending on map rotation.

Given that the detector is able to do its job for the Ace Tournament, and that it works well everywhere else, I suggest we just leave it alone for now. Especially since nothing else uses this detector for now. If we do need better detection at the Blueberry academy in the future, I do have some ideas, such as using a two step waterfill+template match technique. Also, I wrote out the detector's limitations and ideas for future improvement as a comment in the header file.

Reply:
What do the false positives look like?

Reply:
Turquoise colored solid squares. These have a good aspect ratio and area ratio and their rmsd is 80. We can't lower the rmsd threshold because then you false negative at 720p, even for perfect candidates found by the waterfill.

I believe this is because to select for the fly target at 720p at blueberry academy, you need very narrow filters that include your target but not the rest of the blue background. The fly target at 720p has turquoise pixels. 

We only have an issue with getting false positive turquoise squares with narrow filters. With broader filters, it would select more of its surrounding pixels, causing it to fail  the rmsd,  aspect/area ratio checks as expected.

Reply:
If you're talking about the Turquoise square in the center area of the Blueberry map, that looks nothing like the fly icon. So it feels like something else is wrong. I'd try doing a side-by-side comparison of the cropped object along with the actual fly icon.

Reply:
I pushed a new change. It now detects the white, winged portion of the fast travel icon. It seems to work much better. It works even at the blueberry academy at 720p, and doesn't false positive against the winged icon at Pokemon centers (sadly, it still doesn't work indoors at the blueberry academy, due to obstruction from the radar dot overlapping with the fast travel icon. Although, this could be fixed with a custom fast travel icon template.). I was able to get it to work with a single filter, without much need for fine tuning, so I'm hoping it will also be robust against different capture card settings.